Document common terms and vulnerability types (#8125)
jenkins-infra/jenkins.io: content/security/vulnerabilities.adoc

---
layout: security
title: "Vulnerabilities and Scoring"
section: security
---

This page collects information about historically common kinds of security vulnerabilities in Jenkins, including how we score them.
For vulnerability scoring, we use https://www.first.org/cvss/v3-1/[CVSS v3.1].

[#xss]
== Cross-Site Scripting (XSS)

Attackers can inject JavaScript into Jenkins in a way that causes it to be executed in another user's (the "victim") browser when they view certain pages.
https://developer.mozilla.org/en-US/docs/Web/Security/Attacks/XSS[Learn more].

Who can exploit this vulnerability depends on whether it is _stored_ or _reflected_.

> The latter can generally be exploited by anyone, usually by having the victim click a link (CVSS v3.1: `PR:N/UI:R`), while the former usually needs some permission, e.g., View/Configure, to store the payload (CVSS v3.1: `PR:L/UI:R` or `PR:H/UI:R`).

Reviewer: ```suggestion
The latter can generally be exploited by anyone, usually by having the victim click a link (CVSS v3.1: `PR:N/UI:R`), while the former usually needs some permission, such as View/Configure, to store the payload (CVSS v3.1: `PR:L/UI:R` or `PR:H/UI:R`).
```
Suggestion to replace `e.g.` with `such as` per [microsoft style guide](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/eg)

In Jenkins, if the victim is an administrator, features like the link:/doc/book/managing/script-console/[Script Console] allow executing arbitrary code on the controller.
As a result, the impact can be a complete compromise of the Jenkins instance (CVSS v3.1: `C:H/I:H/A:H`).

> A less severe variant of this is when HTML elements (e.g., formatting) can be injected in unexpected places, potentially causing user confusion, but JavaScript execution is prohibited through existing protections.

Reviewer: ```suggestion
A less severe variant of this is when HTML elements, like formatting, can be injected in unexpected places, potentially causing user confusion, but JavaScript execution is prohibited through existing protections.
```
Suggestion to replace `e.g.` with `like` per [microsoft style guide](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/eg).

    ** The attacker needs an API token for a (non-anonymous) user account.
        It is not necessary for this user account to (still) have Overall/Read permission.
* **Remote code execution via "Remember me" cookie** +
    Forging a "Remember me" cookie allows attackers to log in to Jenkins using a web browser, thereby gaining access to the Script Console if they forge a cookie for an administrator account.
    Exploitation requires that all of the following conditions are met:
    ** The "Remember me" feature is enabled (the default).
    ** Attackers have Overall/Read permission to be able to read content in files beyond the first few lines.
* **Remote code execution via stored cross-site scripting (XSS) attacks through build logs** +
    Forging serialized console note objects allows implementing XSS attacks by injecting arbitrary HTML and JavaScript into build logs.
    This attack bypasses the protections added for link:/security/advisory/2017-02-01/#persisted-cross-site-scripting-vulnerability-in-console-notes[SECURITY-382 in the 2017-02-01 security advisory].

>     Exploitation requires that attackers can control build log output (e.g., through pull requests).

Reviewer: ```suggestion
    Exploitation requires that attackers can control build log output, such as through pull requests.
```
Suggestion to replace `e.g.` with `such as` per [microsoft style guide](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/eg)

Because the crumb is temporary and specific to a user's session, it is not accessible to attackers and therefore absent from requests initiated by external websites.

There are two common variants of CSRF vulnerabilities in Jenkins:

* Some actions do not require POST requests, even if the UI for them always sends POST requests anyway.
  As a result, these actions accept GET requests and the CSRF protection for POST requests can easily be bypassed.
* Jenkins allows plugin and extension developers to disable CSRF protection for specific URLs.
  If this is implemented improperly, an unintended consequence of that is the unexpected circumvention of CSRF protection.

[#pt]
== Path Traversal

Path traversal vulnerabilities can occur when file system operations accept user input that isn't sanitized or validated.
If a file with a user-specified name is supposed to be stored in a specific directory, but that name is `../../../config.xml`, one of Jenkins's configuration files may be replaced.

Regardless of whether the path traversal vulnerability affects reading from files or writing to files, the impact can be a complete compromise in both cases (CVSS v3.1: `C:H/I:H/A:H`).

> === Writing Files

Author: Looks like the relative font sizes are a poor choice then…?
